File: leetcode_ly/c++/Stack/DecodeString.cpp
class Solution {
public:
    string decodeString(string s) {
        if(s.size() == 0) return "";
        stack<char> s1;
        for(int i = 0; i < s.size(); i++){
            if(s[i] != ']'){
                s1.push(s[i]);
            }
            else{
                string res = "";
                while(s1.top() != '['){
                    res += s1.top();
                    s1.pop();
                }
                reverse(res.begin(),res.end());
                s1.pop();
                string num = "";
                while(!s1.empty() && s1.top() >= '0' && s1.top() <= '9'){
                    num += s1.top();
                    s1.pop();
                }
                reverse(num.begin(),num.end());
                int cnt = atoi(num.c_str());
                string temp = res;
                for(int j = 0; j < cnt - 1; j++){
                    res += temp;;
                }
                //s1.pop();
                for(int k = 0; k < res.size(); k++){
                    s1.push(res[k]);
                }
            }
        }
        string ans = "";
        while(!s1.empty()){
            ans += s1.top();
            s1.pop();
        }
        reverse(ans.begin(),ans.end());
        return ans;
    }
};

File: leetcode_ly/java/Stack/DecodeString.java
public class Solution {
    public String decodeString(String s) {
        StringBuffer res = new StringBuffer();
        Stack<Character> stack = new Stack<Character>();
        int i = 0;
        while(i < s.length()){
            if(s.charAt(i) != ']') stack.push(s.charAt(i));
            else{
                //获取被重复字符串
                StringBuffer cur = new StringBuffer();
                while(stack.peek() != '['){
                    cur = cur.append(stack.pop());
                }
                cur = cur.reverse();
                stack.pop();
                //获取重复的次数
                StringBuffer digit = new StringBuffer();
                while(!stack.isEmpty() && Character.isDigit(stack.peek())){
                    digit = digit.append(stack.pop());
                }
                digit = digit.reverse();
                int cnt = Integer.parseInt(digit.toString());
                //连接然后入栈
                StringBuffer str = new StringBuffer();
                while(cnt != 0){
                    str = str.append(cur);
                    cnt--;
                }
                for(int j = 0; j < str.length(); j++){
                    stack.push(str.charAt(j));
                }
            }
            i++;
        }
        while(!stack.isEmpty()){
            res = res.append(stack.pop());
        }
        return res.reverse().toString();
    }
}